Srpg 40 character growth

jhwoo944 opened the pull request:
## 개요
<!---- 변경 사항 및 관련 이슈에 대해 간단하게 작성해주세요. 어떻게 보다 무엇을 왜 수정했는지 설명해주세요. -->
ClassSO / TraitSO
- 캐릭터 클래스 / 특성 관련 폴더 / 스크립트 / SO 생성 및 추가.
- 해당 SO 스크립트에 기본 필드 및 스탯 증가값 필드 추가

ChracterSO
- 성장 스탯값 필드 추가.
- ClassSO / TraitSO 필드 추가.
- 성급 필드 추가.

Chracter
- 레벨업 기능 추가 / 경험치, 최대 경험치 및 레벨 관련 데이터 처리 기능 추가.
- 기본 스탯에 성장 스탯을 적용 시키는 메서드 추가.
- 클래스 / 특성 선택 관련 필드 및 메서드 추가.
- 선택한 클래스 / 특성에 의한 증가 스탯 값을 계산하고 저장하는 메서드 추가.
- 캐릭터 성급 / 돌파 관련 필드 및 메서드 추가.
- 치명타 확률, 치명타 피해, 주는 피해 증가, 받는 피해 감소 필드 추가. 

Constants
- 캐릭터 돌파 / 특성 / 클래스 관련 enum 값 추가

자세한 내용은 커밋 메세지를 참고해주세요.
이 구조 설계 및 데이터 처리 방식 / 코드 형식이 적절하고 효율적인지 확신할 수 없으므로, 
자유롭게 피드백 남겨 주세요.
<!---- Resolves: #(Isuue Number) -->

## PR 유형
어떤 변경 사항이 있나요?

- [x] 새로운 기능 추가
- [ ] 버그 수정
- [ ] CSS 등 사용자 UI 디자인 변경
- [ ] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [ ] 코드 리팩토링
- [x] 주석 추가 및 수정
- [ ] 문서 수정
- [ ] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## PR Checklist
PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다. (ex. feature/SRPG-11-UIBase)
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

jhwoo944 committed:
SRPG-23 feat: 캐릭터 성장 능력치 추가 외

-CharacterSO에 GrowStatus 추가
-공격타입 Mage와 Res 능력치는 더 이상 사용하지 않으므로 삭제
-ApplyGrowStat(  ) 메서드 내용 추가
//원본캐릭터의 기본 스탯 + (원본 캐릭터의 성장 스탯 * 이 캐릭터 개체의 현재 레벨)
-Constants에 Character 관련 상수값 enum 추가 (현재 작성된 상수: 캐릭터 기본 치명타확률과 기본 치명타 피해)

jhwoo944 committed:
SRPG-39 feat: 스킬 레벨 추가 외

-Characters.cs에 캐릭터 스킬 레벨 필드 추가
-SkillSO.cs에 growCoefficient; //(레벨 당 증가하는 스킬 계수) 필드 추가
-스킬 데미지 타입 (물리데미지/마법데미지)는 더 이상 사용하지 않으므로 주석처리함.
-이후 캐릭터 성장 및 성장치 관련 작업은 브랜치 나눠서 작업

jhwoo944 committed:
SRPG-40 Feat: 캐릭터 클래스 및 특성 SO 스크립트 생성

-특성과 클래스 정보를 저장하는 SO 폴더와 스크립트를 생성했습니다.
-스크립트는 아직 미작성 상태입니다.

jhwoo944 committed:
SRPG-40 Feat: TraitSO.cs 작성 및 스크립터블 오브젝트 생성

TraitSO.cs 작성 및 스크립터블 오브젝트 생성

jhwoo944 committed:
SRPG-40 Feat:특성SO 스크립트 내용추가

-스탯 증가 합연산 수치 및 곱연산 수치 필드 추가

jhwoo944 committed:
Merge remote-tracking branch 'origin/develop' into SRPG-40-CharacterGrowth

jhwoo944 committed:
Merge remote-tracking branch 'origin/develop' into SRPG-40-CharacterGrowth

jhwoo944 committed:
SRPG-40 Feat: Character에 클래스와 특성 필드를 입력

-Character에 클래스와 특성, 성급 필드를 입력함.
-스탯 계산은 아직 구현되지 않음.

jhwoo944 committed:
Merge remote-tracking branch 'origin/develop' into SRPG-40-CharacterGrowth

jhwoo944 committed:
SRPG-40 Feat: 특성 배율 증가 초기값 1로 설정

-특성 및 클래스의 스탯 배율 증가 필드의 데이터 형식을 float로 바꾸고 초기값을 1로 변경함. (계산식에 적용시킬 때 간편하게 만들기 위해)

jhwoo944 committed:
SRPG-40 Fix: 경로 변경 및 스크립트 작업

-특성과 클래스 스크립트 폴더의 경로 캐릭터 폴더 내부로 이동
-주석 추가

jhwoo944 committed:
Merge remote-tracking branch 'origin/develop' into SRPG-40-CharacterGrowth

jhwoo944 committed:
SRPG-40 Feat: 캐릭터 베이스, 성장 및 능력치 관련 작업

- 캐릭터 클래스 및 특성 관련 폴더, 스크립트, 스크립터블 오브젝트 생성 및 추가.
- 클래스/특성 SO는 Skill과 유사한 구조로 CharacterSO에서 설정할 수 있습니다.
- Chracter.에서 클래스와 특성를 선택하는 메서드를 추가했습니다.
- 클래스/특성 SO에서 캐릭터가 해당 특성/클래스 설정 시 증가 시킬 Atk, Def, Health, Mov, 치명타 확률, 치명타 피해, 주는 피해 증가, 받는 피해 감소 스탯 값을 저장할 수 있습니다.
- Chracter에서 선택한 클래스/특성에 저장된 증가 스탯 값을 CharacterSO를 통해 불러온 뒤 기본 스탯을 베이스로 합연산/곱연산을 계산하고, Calc 스탯값에 저장하는 메서드를 추가했습니다.
- 현재 특성/ 클래스는 스탯 능력치 증가값만 유효하고, 그 외에 추가로 적용할 수 있는 특수 기능에 대한 로직은 미구현 상태입니다. (구현 필요)
- 캐릭터 경험치 및 레벨업 기능, 레벨업 관련 데이터 처리 기능을 추가했습니다.
- ChracterSO에 성장 스탯 증가 값을 지정할 수 있도록 필드를 추가하고
레벨업 시 (기본 스탯 = 기본스탯 + (level*성장 스탯 증가값))을 저장하는 메서드를 추가했습니다.
- 캐릭터 돌파 / 특성 / 클래스 관련 enum 값을 Constants에 추가했습니다.
- 캐릭터 돌파 메서드를 추가했으나, 조각 아이템과 상호작용은 미구현 상태입니다. (추후 구현 필요)
- Chracter에 치명타 확률, 치명타 피해, 주는 피해 증가, 받는 피해 감소 필드를 추가했습니다. 실제 인게임 전투에서 적용되도록 추후 작업이 필요합니다.
- 이 구조 설계 및 데이터 처리 방식/ 코드 형식이 적절하고 효율적인지 확신할 수 없으므로 자유롭게 피드백 남겨 주세요.

jhwoo944 committed:
Merge remote-tracking branch 'origin/develop' into SRPG-40-CharacterGrowth

GeCrabKani reviewed (approved):
SkillSO에 스킬 레벨 변수가 없습니다.

jhwoo944 commented:
> SkillSO에 스킬 레벨 변수가 없습니다.

추가했었는데 Develop에서 merge하는 과정에서 누락된 것 같네요
추가하겠습니다.

jhwoo944 committed:
SRPG-40 Fix: 오타 수정 / 스킬 레벨 추가

- 일부 오타 수정
- Character의 일부 필드에 자동 프로퍼티 ( private set ) 추가
- Character에 스킬 / 패시브 스킬 레벨 필드 추가
- 데미지 타입 (물리/마법)은 더 이상 사용하지 않기로 했으므로 주석처리함
- 주석 추가

jhwoo944 committed:
SRPG-40 Feat : 캐릭터 레벨 조건 추가

- 캐릭터 레벨이 계정 레벨을 넘지 못하도록 조건 추가함

jhwoo944 committed:
SRPG-40 Feat: 캐릭터 장비 폴더 생성 및 스크립트 작성 중

-Equip 폴더 생성
-Weapon과 Armor SO를 분리해서 작성하는 중

jhwoo944 committed:
SRPG-40 CharacterGrowth 작업

CharacterGrowth 클래스를 새로 생성하고. Character에서 다음의 요소를 분리했습니다.
 - 필드 : 레벨 / 경험치 / 특성 / 클래스 / 성급 / 스킬 레벨 / 장비 레벨 / 호감도 레벨
 - 메서드 : 성장 스탯 계산 메서드 / 추가 스탯 계산 메서드 / 특성, 클래스 선택 메서드 / 돌파 메서드 / 스킬, 장비, 호감도 레벨업 메서드

 - 스탯 관련 : Character가 기본 스탯과 Calc 스탯을 가지고 있고,
CharacterGrowth의 스탯 계산 메서드에서 스탯을 계산한 뒤 Character의 기본 스탯과 Calc 스탯에 저장합니다.

 Constants.cs
 - 장비 타입, 무기 종류, 방어구 종류 enum 추가
